tomcat/src/main/java/com/linecorp/armeria/server/tomcat/TomcatService.java

```diff
@@ -254,125 +240,19 @@ static String toString(org.apache.catalina.Server server) {
-        if (hostname == null) {
-            hostname = cfg.server().defaultHostname();
-        }
-
-        if (armeriaServer != null) {
-            if (armeriaServer != cfg.server()) {
-                throw new IllegalStateException("cannot be added to more than one server");
-            } else {
-                return;
-            }
-        }
-
-        armeriaServer = cfg.server();
-        armeriaServer.addListener(configurator);
-    }
-
     /**
      * Returns Tomcat {@link Connector}.
      */
-    public Connector connector() {
-        final Connector connector = this.connector;
-        if (connector == null) {
-            throw new IllegalStateException("not started yet");
-        }
-
-        return connector;
-    }
-
-    void start() throws Exception {
-        assert hostname != null;
-        started = false;
-        connector = connectorFactory.apply(hostname);
-        final Service service = connector.getService();
-        if (service == null) {
-            return;
-        }
-
-        final Engine engine = TomcatUtil.engine(service, hostname);
-        final String engineName = engine.getName();
-        if (engineName == null) {
-            return;
-        }
-
-        if (activeEngines.contains(engineName)) {
-            throw new TomcatServiceException("duplicate engine name: " + engineName);
-        }
-
-        server = service.getServer();
-
-        if (!TOMCAT_START_STATES.contains(server.getState())) {
-            logger.info("Starting an embedded Tomcat: {}", toString(server));
-            server.start();
-            started = true;
-        }
-
-        activeEngines.add(engineName);
-        this.engineName = engineName;
-    }
-
-    void stop() throws Exception {
-        final org.apache.catalina.Server server = this.server;
-        final Connector connector = this.connector;
-        this.server = null;
-        this.connector = null;
+    public abstract Optional<Connector> connector();
 
-        if (engineName != null) {
-            activeEngines.remove(engineName);
-            engineName = null;
-        }
-
-        if (server == null || !started) {
-            return;
-        }
-
-        try {
-            logger.info("Stopping an embedded Tomcat: {}", toString(server));
-            server.stop();
-        } catch (Exception e) {
-            logger.warn("Failed to stop an embedded Tomcat: {}", toString(server), e);
-        }
-
-        postStopTask.accept(connector);
-    }
+    @Nullable
+    abstract String hostName();
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        final Adapter coyoteAdapter = connector().getProtocolHandler().getAdapter();
-        if (coyoteAdapter == null) {
+        final Optional<Adapter> coyoteAdapter = connector().map(Connector::getProtocolHandler)
+                                                           .map(ProtocolHandler::getAdapter);
```

Comment:
Also need to condense this one to `connector().map(c -> c.getProtocolHandler().getAdapter())`